File: core/src/main/java/heast/core/network/packets/c2s/auth/LogoutC2SPacket.java
package heast.core.network.packets.c2s.auth;

import heast.core.network.Packet;
import heast.core.network.PacketBuf;
import heast.core.network.listeners.ServerAuthListener;

public final class LogoutC2SPacket implements Packet<ServerAuthListener> {

    public LogoutC2SPacket() {

    }

    public LogoutC2SPacket(PacketBuf buf) {

    }

    @Override
    public void write(PacketBuf buf) {

    }

    @Override
    public void apply(ServerAuthListener listener) {
        listener.onLogout();
    }
}

File: Core/Network/Packets/C2S/LogoutC2SPacket.cs
using Core.Network.Listeners;

namespace Core.Network.Packets.C2S;

public partial class LogoutC2SPacket : AbstractPacket
{
    public enum LogoutReason
    {
        Logout,
        Quit,
        Kick,
        Ban
    }

    public LogoutC2SPacket(LogoutReason reason)
    {
        Reason = reason;
    }

    public LogoutReason Reason { get; }

    public override void Apply(IPacketListener listener)
    {
        if (listener is IServerAuthListener authListener)
        {
            authListener.OnLogout(this);
        }
    }
}